planning/behavior_velocity_planner/autoware_behavior_velocity_blind_spot_module/src/time_to_collision.cpp

```diff
@@ -0,0 +1,239 @@
+  }
+
+  return std::make_optional<std::pair<double, double>>(entry_time.value(), exit_time.value());
+}
+
+std::vector<std::tuple<double, double, autoware_perception_msgs::msg::PredictedPath>>
+compute_time_interval_for_passing_line(
+  const autoware_perception_msgs::msg::PredictedObject & object,
+  const lanelet::ConstLineString3d & line1, const lanelet::ConstLineString3d & line2)
+{
+  std::vector<std::tuple<double, double, autoware_perception_msgs::msg::PredictedPath>>
+    passage_time_intervals;
+  for (const auto & predicted_path : object.kinematics.predicted_paths) {
+    if (predicted_path.path.size() < 2) {
+      continue;
+    }
+    const double time_step = predicted_path.time_step.sec + predicted_path.time_step.nanosec * 1e-9;
+    const double horizon = time_step * static_cast<double>(predicted_path.path.size());
+    static constexpr double new_time_step = 0.1;
+    const auto precise_predicted_path = autoware::object_recognition_utils::resamplePredictedPath(
+      predicted_path, new_time_step, horizon);
+    const auto & shape = object.shape;
+
+    // search forward
+    std::optional<double> entry_time{};
+    for (const auto & [i, pose] : ranges::views::enumerate(precise_predicted_path.path)) {
+      const auto object_poly = autoware_utils_geometry::to_polygon2d(pose, shape);
+      if (boost::geometry::intersects(object_poly, lanelet::utils::to2D(line1).basicLineString())) {
+        entry_time = i * new_time_step;
+        break;
+      }
+    }
+
+    // search backward
+    std::optional<double> exit_time{};
+    for (const auto & [i, pose] :
+         ranges::views::enumerate(precise_predicted_path.path | ranges::views::reverse)) {
+      const auto object_poly = autoware_utils_geometry::to_polygon2d(pose, shape);
+      const double time = horizon - i * new_time_step;
+      if (entry_time && time < entry_time.value()) {
+        break;
+      }
+      if (boost::geometry::intersects(object_poly, lanelet::utils::to2D(line2).basicLineString())) {
+        exit_time = time;
+        break;
+      }
+    }
+    if (!exit_time) {
+      continue;
+    }
+    // in case the object is completely inside conflict_area, it is regarded entry_time = 0.0
+    passage_time_intervals.emplace_back(
+      entry_time.value_or(0.0), exit_time.value(), predicted_path);
+  }
+  return passage_time_intervals;
+}
+
+autoware_internal_planning_msgs::msg::SafetyFactor UnsafeObject::to_safety_factor() const
+{
+  autoware_internal_planning_msgs::msg::SafetyFactor factor;
+  factor.type = autoware_internal_planning_msgs::msg::SafetyFactor::OBJECT;
+  factor.object_id = object.object_id;
+  factor.predicted_path = predicted_path;
+  factor.ttc_begin = std::get<0>(object_passage_interval);
+  factor.ttc_end = std::get<1>(object_passage_interval);
+  factor.points.push_back(object.kinematics.initial_pose_with_covariance.pose.position);
+  factor.is_safe = false;
+  return factor;
+}
+
+std::optional<double> is_critical_for_ego(
```

Comment:
the name of the function can be improved. Maybe `get_unsafe_time` ?

Reply:
addressed in f3d3f9802c8a7424d5775c7e7155f65633b8dc83